File: reactor/r2dbc-mysql/src/main/java/ua/mai/zyme/r2dbcmysql/exception/FaultException.java
package ua.mai.zyme.r2dbcmysql.exception;

import org.springframework.http.HttpStatus;

import java.util.List;

public class FaultException extends RuntimeException {

    private FaultDetails faultDetails;

    public FaultException(FaultInfo faultInfo) {
        this(faultInfo, List.of());
    }

    public FaultException(FaultInfo faultInfo, Object...errorParameters ) {
        this(null, faultInfo, errorParameters);
    }

    public FaultException(Throwable cause, FaultInfo faultInfo, Object...errorParameters) {
        super(cause);
        faultDetails = new FaultDetails(faultInfo, errorParameters);
    }

    public FaultException(Throwable cause) {
        super(cause);
        faultDetails = new FaultDetails(AppFaultInfo.UNEXPECTED_ERROR, cause.getMessage());
    }


    public String getCode() {
        return faultDetails.faultInfo.code();
    }

    public String getMessage() {
        return getFaultMessage();
    }

    public HttpStatus getHttpStatus() {
        return faultDetails.faultInfo.httpStatus();
    }

    public String getFaultMessage() {
        return faultDetails.getMessage();
    }

    public String getCauseMessage() {
        return getCause() != null ? getCause().getMessage() : null;
    }

    public List<Object> getErrorParameters() {
        return faultDetails.errorParameters;
    }

    public String toString() {
        return "Code=" + getCode() +
               ", Message=\"" + getFaultMessage() + "\"" +
                (getCause() != null ? (", Details=" + getCause()) : "") +
               ", HttpStatus=" + getHttpStatus();
    }

}

File: kotlin/kotlin-r2dbc-mysql/kotlin-r2dbc-mysql-be/src/main/java/ua/mai/zyme/r2dbcmysql/exception/FaultException.kt
package ua.mai.zyme.r2dbcmysql.exception

import org.springframework.http.HttpStatus

class FaultException : RuntimeException {

    private val faultDetails: FaultDetails

    constructor(faultInfo: FaultInfo, vararg errorParameters: Any?) : super() {
        this.faultDetails = FaultDetails(faultInfo, errorParameters.toList())
    }

    constructor(cause: Throwable?, faultInfo: FaultInfo, vararg errorParameters: Any?) : super(cause) {
        this.faultDetails = FaultDetails(faultInfo, errorParameters.toList())
    }

    constructor(cause: Throwable?) : super(cause) {
        this.faultDetails = FaultDetails(AppFaultInfo.UNEXPECTED_ERROR, cause?.message)
    }

    val code: String
        get() = faultDetails.faultInfo.code()

    override val message: String
        get() = faultMessage

    val httpStatus: HttpStatus
        get() = faultDetails.faultInfo.httpStatus()

    val faultMessage: String
        get() = faultDetails.getMessage()

    val causeMessage: String?
        get() = cause?.message

    val errorParameters: List<Any?>
        get() = faultDetails.errorParameters

    override fun toString(): String {
        return "Code=$code, Message=\"$faultMessage\"" +
                (cause?.let { ", Details=$it" } ?: "") +
                ", HttpStatus=$httpStatus"
    }

}